### ISSUE TITLE
graphql dependency is mismatched with graphql-compose

### ISSUE THREAD

**Issue by adambaratz** [gatsbyjs/gatsby#35212]
### Preliminary Checks

- [X] This issue is not a duplicate. Before opening a new issue, please search existing issues: https://github.com/gatsbyjs/gatsby/issues
- [X] This issue is not a question, feature request, RFC, or anything other than a bug report directly related to Gatsby. Please post those things in GitHub Discussions: https://github.com/gatsbyjs/gatsby/discussions

### Description

There is a [known issue](https://github.com/graphql-compose/graphql-compose/issues/374) with graphql-compose where it technically depends on graphql v16, but isn't explicit about it. Gatsby, after #34787, depends on this package but depends on an earlier version of graphql. As a result, the TypeScript compiler fails unless you enable the [`skipLibCheck` option](https://www.typescriptlang.org/tsconfig#skipLibCheck).

### Reproduction Link

N/A

### Steps to Reproduce

1.
2.
3.
...

### Expected Result

No errors with tsc.

### Actual Result

```
$ tsc
node_modules/graphql-compose/lib/type/buffer.d.ts:3:25 - error TS2315: Type 'GraphQLScalarType' is not generic.

3 declare const _default: GraphQLScalarType<Buffer, string>;
                          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

node_modules/graphql-compose/lib/type/date.d.ts:2:25 - error TS2315: Type 'GraphQLScalarType' is not generic.

2 declare const _default: GraphQLScalarType<Date, string>;
                          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


Found 2 errors.
```

### Environment

```shell
.
```

**Comment by adambaratz** [comment]
Apologies if I short-circuited the template in an unhelpful way, but I think the fix is as simple as updating the graphql dependency (assuming that doesn't break anything). If it would be helpful to create a very simple project to be able to reproduce, I can do that.

**Comment by chris48s** [comment]
Commenting because this issue has not been fixed

**Comment by ctmm** [comment]
Also experiencing issues with this topic

**Comment by LekoArts** [comment]
While on paper it's "just" updating `graphql` the breaking changes list is quite long:

https://github.com/graphql/graphql-js/releases/tag/v16.0.0

We'd appreciate any help with putting up a PR updating it and assessing which breaking changes affect us and if this would also mean a breaking change for Gatsby users.

**Comment by ctmm** [comment]
In the meantime, we've downgraded our version of `graphql-composer` – the package triggering the conflict.
On `package.json`:
```json
  "overrides": {
    "gatsby": {
      "graphql-compose": "9.0.4"
    },
    "gatsby-plugin-manifest": {
      "graphql-compose": "9.0.4"
    },
    "gatsby-plugin-utils": {
      "graphql-compose": "9.0.4"
    }
  },
```
Sharing it, as it might be helpful for others.

At the same time, I'm unsure if the issue will be fixed by upgrading Gatsby's graphql version.
https://github.com/graphql-compose/graphql-compose/issues/383
https://github.com/graphql-compose/graphql-compose/issues/374

Should we fix it by downgrading Gatsby's desired version of `graphql-compose`?

**Comment by adambaratz** [comment]
Still an issue.

**Comment by LekoArts** [comment]
Downgrading `graphql-compose` is not an option by the way.
